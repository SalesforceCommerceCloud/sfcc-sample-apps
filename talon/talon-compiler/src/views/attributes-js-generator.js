const beautify = require('js-beautify');
const ROUTE_PARAM_REFERENCE = "routeParams";
const CMP_REFERENCE = "cmp";
const EXPRESSION_REGEX_INNER_CAPTURE = /\{!\s*([a-zA-Z0-9_.]+)\s*\}/;

const DEFAULT_ATTRIBUTES = '(cmp) => ({})';

/**
 * Generates the Javascript for the function that takes a component and returns the attribute set
 * with resolved references to route parameters.
 * e.g. (cmp) => ( { ...attributeSet... } )
 *
 * @param {Object} attributeSet the aggregated component attributes for a view
 */
function generateAttributesJS(attributeSet) {
    const attributeSetString = attributeSetToJS(attributeSet);
    return `(${CMP_REFERENCE}) => (${beautify(attributeSetString)})`;
}

/**
 * Converts the attribute set of a view to Javascript.
 *
 * @param {Object} attributeSet the aggregated component attributes for a view
 */
function attributeSetToJS(attributeSet) {
    let attributeSetString = '{';
    let cmpIdx = 0;
    Object.entries(attributeSet).forEach(([cmpNameKey, cmpAttributes]) => {
        // Exclude empty component attribute maps
        if (Object.keys(cmpAttributes).length === 0) {
            return;
        }
        if (cmpIdx++ > 0) {
            attributeSetString += ',';
        }
        attributeSetString += `"${cmpNameKey}": {`;
        let attrIdx = 0;
        Object.entries(cmpAttributes).forEach(([attributeName, attributeValue]) => {
            if (attrIdx++ > 0) {
                attributeSetString += ',';
            }
            attributeSetString += `"${attributeName}":${attributeValueToJS(attributeValue)}`;
        });
        attributeSetString += `}`;
    });
    attributeSetString += '}';
    return attributeSetString;
}

/**
 * Convert an attribute value to Javascript.
 * The value is resolved as necessary in the Javascript.
 *
 * @param {*} attributeValue value of the attribute
 */
function attributeValueToJS(attributeValue) {
    if (typeof attributeValue === "string" && attributeValue.match(EXPRESSION_REGEX_INNER_CAPTURE)) {
        // Resolve any {!EL}
        return resolveStringAttribute(attributeValue);
    }
    return JSON.stringify(attributeValue);
}

/**
 * Evaluates if a string attribute contains any {!expression} and replaces any instances with a reference to route param(s).
 * If the param is not present, the value defaults to an empty string.
 * e.g. "prefix {!recordId} postfix" => "prefix " + (cmp.routeParams["recordId"] || '') + " postfix"
 *
 * Note: This assumes that values from cmp.routeParams will always evaluate to strings (and never numbers).
 *
 * @param {string} attributeValue string attribute value containing an {!expression}
 */
function resolveStringAttribute(attributeValue) {
    let remainderOfString = attributeValue;
    let match = EXPRESSION_REGEX_INNER_CAPTURE.exec(remainderOfString);
    const segments = [];
    while (match) {
        const prefix = remainderOfString.substring(0, match.index);
        if (prefix) {
            segments.push(JSON.stringify(prefix));
        }
        const exprKey = match[1].trim(); // capture group, e.g. "recordId" from "{! recordId }"
        // Default to empty string if route param not present
        const routeParamRef = `(${CMP_REFERENCE}.${ROUTE_PARAM_REFERENCE}["${exprKey}"] || '')`;
        segments.push(routeParamRef);
        // Check for more matches
        remainderOfString = remainderOfString.substring(match.index + match[0].length);
        match = EXPRESSION_REGEX_INNER_CAPTURE.exec(remainderOfString);
    }
    if (remainderOfString) {
        segments.push(JSON.stringify(remainderOfString));
    }
    return segments.join(' + ');
}

module.exports = { generateAttributesJS, DEFAULT_ATTRIBUTES };