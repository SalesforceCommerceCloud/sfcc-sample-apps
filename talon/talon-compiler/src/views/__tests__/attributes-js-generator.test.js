import { generateAttributesJS } from '../attributes-js-generator.js';

const attributeLiterals = [
    null,
    true,
    false,
    -1.5,
    3.14,
    420,
    "",
    "foo",
    'escaped \'single\' quotes',
    "escaped \"double\" quotes",
    "\\",
    {},
    {
        notEvaluated: "{!recordId}"
    },
    {
        a : "b",
        c : {
            d : true
        },
        e : 3.14,
        "f g" : {}
    },
    [],
    [1],
    [1, false, null, "{!ignored}", [], {}, [3.14]],
    "{ !expressionsStartWithCurlyBang }",
    "{! hyphens-not-allowed }",
    "{!}",
    "{!noEndingBrace"
];
const attributeExpressions = [
    "{!recordId}",
    "{!   whitespacePrefixPostfixAllowed   }",
    "{!underscores_are_allowed}",
    "{!periods.are.allowed}",
    "{! 09812309123 }",
    "{! 09_.1enmixed0_.aSS }"
];
const attributeEmbeddedExpressions = [
    "   {!recordId}   ",
    "prefix {!recordId}",
    "{!recordId} postfix",
    "prefix{!recordId}postfix",
    "a{!recordId}b{!userId}c{!recordId}d",
    "{!recordId}{!userId}",
    " \"escaped\" \"{!recordId}\"",
    "{!{!recordId}}",
    "{!{!recor}dId}",
];
const defaultAttributeSet = {
    "xliterals_0": {
        str: "foo",
        bool: true,
    },
    "xexpressions_1": {
        recordId: "{!recordId}",
        userId: "{!myValue}"
    },
    "xmixed_2": {
        literal: "literal",
        exprAttr: "{!retVal}",
        literal2: "literal2",
    },
    "xsingle-expression_3": {
        userId: "{!myValue}"
    },
    "xsingle-literal_4": {
        one: "attribute"
    },
    "xempty_5": {},
    "xembedded-expressions_6": {
        complexAttr: "prefix {!recordId} postfix"
    },
};

// Create a simple single-component attribute set from a list of values
function createAttributeSet(valueList) {
    const cmpNameKey = "xfoo_0";
    const attributeSet = {};
    attributeSet[cmpNameKey] = {};
    for (let i = 0; i < valueList.length; i++) {
        attributeSet[cmpNameKey][i] = valueList[i];
    }
    return attributeSet;
}

describe("attributes-js-generator", () => {
    describe("generateAttributesJS", () => {
        describe("generates the right javascript for", () => {
            it("empty attribute set", () => {
                const generatedJS = generateAttributesJS({});
                expect(generatedJS).toMatchSnapshot();
            });
            it("attribute set with empty component maps", () => {
                const generatedJS = generateAttributesJS({
                    "xempty_1": {},
                    "xempty_2": {},
                    "xempty_3": {},
                });
                expect(generatedJS).toMatchSnapshot();
            });
            it("attribute set with literals", () => {
                const attributeSet = createAttributeSet(attributeLiterals);
                const generatedJS = generateAttributesJS(attributeSet);
                expect(generatedJS).toMatchSnapshot();
            });
            it("attribute set with {!expressions}", () => {
                const attributeSet = createAttributeSet(attributeExpressions);
                const generatedJS = generateAttributesJS(attributeSet);
                expect(generatedJS).toMatchSnapshot();
            });
            it("attribute set with embedded {!expressions}", () => {
                const attributeSet = createAttributeSet(attributeEmbeddedExpressions);
                const generatedJS = generateAttributesJS(attributeSet);
                expect(generatedJS).toMatchSnapshot();
            });
            it("mixed attribute set", () => {
                const generatedJS = generateAttributesJS(defaultAttributeSet);
                expect(generatedJS).toMatchSnapshot();
            });
        });
    });
});