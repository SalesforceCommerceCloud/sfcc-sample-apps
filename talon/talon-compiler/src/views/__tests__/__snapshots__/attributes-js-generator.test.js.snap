// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributes-js-generator generateAttributesJS generates the right javascript for attribute set with {!expressions} 1`] = `
"(cmp) => ({
    \\"xfoo_0\\": {
        \\"0\\": (cmp.routeParams[\\"recordId\\"] || ''),
        \\"1\\": (cmp.routeParams[\\"whitespacePrefixPostfixAllowed\\"] || ''),
        \\"2\\": (cmp.routeParams[\\"underscores_are_allowed\\"] || ''),
        \\"3\\": (cmp.routeParams[\\"periods.are.allowed\\"] || ''),
        \\"4\\": (cmp.routeParams[\\"09812309123\\"] || ''),
        \\"5\\": (cmp.routeParams[\\"09_.1enmixed0_.aSS\\"] || '')
    }
})"
`;

exports[`attributes-js-generator generateAttributesJS generates the right javascript for attribute set with embedded {!expressions} 1`] = `
"(cmp) => ({
    \\"xfoo_0\\": {
        \\"0\\": \\"   \\" + (cmp.routeParams[\\"recordId\\"] || '') + \\"   \\",
        \\"1\\": \\"prefix \\" + (cmp.routeParams[\\"recordId\\"] || ''),
        \\"2\\": (cmp.routeParams[\\"recordId\\"] || '') + \\" postfix\\",
        \\"3\\": \\"prefix\\" + (cmp.routeParams[\\"recordId\\"] || '') + \\"postfix\\",
        \\"4\\": \\"a\\" + (cmp.routeParams[\\"recordId\\"] || '') + \\"b\\" + (cmp.routeParams[\\"userId\\"] || '') + \\"c\\" + (cmp.routeParams[\\"recordId\\"] || '') + \\"d\\",
        \\"5\\": (cmp.routeParams[\\"recordId\\"] || '') + (cmp.routeParams[\\"userId\\"] || ''),
        \\"6\\": \\" \\\\\\"escaped\\\\\\" \\\\\\"\\" + (cmp.routeParams[\\"recordId\\"] || '') + \\"\\\\\\"\\",
        \\"7\\": \\"{!\\" + (cmp.routeParams[\\"recordId\\"] || '') + \\"}\\",
        \\"8\\": \\"{!\\" + (cmp.routeParams[\\"recor\\"] || '') + \\"dId}\\"
    }
})"
`;

exports[`attributes-js-generator generateAttributesJS generates the right javascript for attribute set with empty component maps 1`] = `"(cmp) => ({})"`;

exports[`attributes-js-generator generateAttributesJS generates the right javascript for attribute set with literals 1`] = `
"(cmp) => ({
    \\"xfoo_0\\": {
        \\"0\\": null,
        \\"1\\": true,
        \\"2\\": false,
        \\"3\\": -1.5,
        \\"4\\": 3.14,
        \\"5\\": 420,
        \\"6\\": \\"\\",
        \\"7\\": \\"foo\\",
        \\"8\\": \\"escaped 'single' quotes\\",
        \\"9\\": \\"escaped \\\\\\"double\\\\\\" quotes\\",
        \\"10\\": \\"\\\\\\\\\\",
        \\"11\\": {},
        \\"12\\": {
            \\"notEvaluated\\": \\"{!recordId}\\"
        },
        \\"13\\": {
            \\"a\\": \\"b\\",
            \\"c\\": {
                \\"d\\": true
            },
            \\"e\\": 3.14,
            \\"f g\\": {}
        },
        \\"14\\": [],
        \\"15\\": [1],
        \\"16\\": [1, false, null, \\"{!ignored}\\", [], {},
            [3.14]
        ],
        \\"17\\": \\"{ !expressionsStartWithCurlyBang }\\",
        \\"18\\": \\"{! hyphens-not-allowed }\\",
        \\"19\\": \\"{!}\\",
        \\"20\\": \\"{!noEndingBrace\\"
    }
})"
`;

exports[`attributes-js-generator generateAttributesJS generates the right javascript for empty attribute set 1`] = `"(cmp) => ({})"`;

exports[`attributes-js-generator generateAttributesJS generates the right javascript for mixed attribute set 1`] = `
"(cmp) => ({
    \\"xliterals_0\\": {
        \\"str\\": \\"foo\\",
        \\"bool\\": true
    },
    \\"xexpressions_1\\": {
        \\"recordId\\": (cmp.routeParams[\\"recordId\\"] || ''),
        \\"userId\\": (cmp.routeParams[\\"myValue\\"] || '')
    },
    \\"xmixed_2\\": {
        \\"literal\\": \\"literal\\",
        \\"exprAttr\\": (cmp.routeParams[\\"retVal\\"] || ''),
        \\"literal2\\": \\"literal2\\"
    },
    \\"xsingle-expression_3\\": {
        \\"userId\\": (cmp.routeParams[\\"myValue\\"] || '')
    },
    \\"xsingle-literal_4\\": {
        \\"one\\": \\"attribute\\"
    },
    \\"xembedded-expressions_6\\": {
        \\"complexAttr\\": \\"prefix \\" + (cmp.routeParams[\\"recordId\\"] || '') + \\" postfix\\"
    }
})"
`;
